ivy-prescient.el

```diff
@@ -202,70 +143,45 @@ This is the value that was associated to
 (defvar ivy-prescient--old-initial-inputs-alist nil
   "Previous value of `ivy-initial-inputs-alist'.")
 
-(defun ivy-prescient--wrap-action (caller action)
+(declare-function ivy-state-sort "ivy")
+(declare-function ivy-state-collection "ivy")
+(declare-function ivy--sort-function "ivy")
+(declare-function ivy--get-action "ivy")
+
+(defun ivy-prescient--wrap-action (action)
   "Wrap an action for use in `ivy-read'.
-CALLER is the `:caller' argument to `ivy-read', and ACTION is the
-original action, a function. Return a new function that also
-invokes `prescient-remember'."
-  (lambda (result)
-    ;; Same as in `ivy-prescient-sort-function', we have to account
-    ;; for candidates which are lists by taking their cars. Make sure
-    ;; to do this only for the call to `prescient-remember', and not
-    ;; for the actual action, though. See
-    ;; https://github.com/raxod502/prescient.el/issues/12.
-    (let ((result result))
-      (when (listp result)
-        (setq result (car result)))
-      (unless (memq caller ivy-prescient-excluded-commands)
-        (prescient-remember result)))
-    (when action
-      (funcall action result))))
-
-(cl-defun ivy-prescient-read
-    (ivy-read prompt collection &rest rest &key action caller keymap
-              &allow-other-keys)
-  "Delegate to `ivy-read', handling persistence and sort customization.
-If the `:caller' passed to `ivy-read' is a member of
+ACTION is the original action, a function. Return a new function
+that also invokes `prescient-remember'."
+  (defvar ivy-marked-candidates)
+  (defvar ivy-last)
+  (defvar ivy--directory)
+  (declare-function 'ivy-state-collection "ivy")
+  (if (or ivy-marked-candidates
+          (not (memq (let ((sort (ivy-state-sort ivy-last))
+                           (coll (ivy-state-collection ivy-last)))
+                       (cond ((functionp sort)
+                              sort)
+                             ((or sort (eq coll #'read-file-name-internal))
+                              (ivy--sort-function coll))))
+                     '(ivy-prescient-sort-file-function
+                       ivy-prescient-sort-function))))
+      action
+    (let ((dir ivy--directory))
+      (lambda (x)
+        (let ((cand x))
+          (when (listp cand) (setq cand (car x)))
+          (when dir (setq cand (string-remove-prefix dir cand)))
```

Comment:
I think we should be able to use your suggestion of `file-relative-path`, because if `ivy--directory` is non-nil, then we know that `cand` is a file path. Also I don't think we need to worry about case insensitivity, because `ivy--directory` is appended to `x` before the call to action, and therefore should match exactly.

Reply:
In fact, `file-relative-name` takes care of case-insensitivity for us! :)